examples/extensions/kubernetes.json

```diff
@@ -0,0 +1,38 @@
+{
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes"
+    },
+    "masterProfile": {
+      "count": 1,
+      "dnsPrefix": "",
+      "vmSize": "Standard_D2_v2"
+    },
+    "agentPoolProfiles": [
+      {
+        "name": "agentpool1",
+        "count": 3,
+        "vmSize": "Standard_D2_v2",
+        "availabilityProfile": "AvailabilitySet"
+      }
+    ],
+    "linuxProfile": {
+      "adminUsername": "azureuser",
+      "ssh": {
+        "publicKeys": [
+          {
+            "keyData": ""
+          }
+        ]
+      }
+    },
+    "extensionsProfile": [
+      { "name": "hello-world-k8s", "version": "v1" }
+    ],

```

Comment:
The extensions are really just taking advantage of custom script extensions and linked templates.  What runs where is really up to the extension author.  That is really the beauty of this model - whatever you can do in ARM is supported.

For example, in hello-world-k8s, looking at template-link.json (this is the json that gets injected in the ARM template that links to the linked template), we chose to pass a parameter called masterVMNamePrefix.  In there we passed the name of the 1st master VM. 
` "masterVMNamePrefix": {
                "value": "[concat(variables('masterVMNamePrefix'), '0')]"
},` 
In the template.json, we take in that parameter and set the name of the extension to: 
`"name": "[concat(parameters('masterVMNamePrefix'), '/cse0')]"`

Because the name prior to the '/' matches the name of the first master VM, the script will be run on the first master.

However, you have full flexibility.  To run on all masters or on all agents in a pool, you would pass a prefix in template-link.json and run a loop in template.json.

To run on all masters and on all agents in a pool, you could pass a masterPrefix, masterCount, agentPool1Prefix, agentPool1Count and have 2 separate loops in template.json.

Whatever options are available in ARM are available in the extensions.